### PR TITLE
Implement Barclay documentation framework

### LINKED ISSUE #182: Use Barclay documentation generation
Instead of the Wiki, we should use a gh-pages branch to store the documentation, and generate it automatically with Barclay. For that, we should learn how the system works.

### PR DESCRIPTION
**Pull request by magicDGS**
Close #182

* Add basic Barclay documentation framework
* Jekyll templates
* Annotate documented features
* Make public argument collection to allow their use in the doc generation

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -155,3 +167,44 @@ shadowJar {
     mergeServiceFiles()
     finalizedBy currentJar
 }
+
+// Generate ReadTools Online Doc files (jekyll formatted)
+task readtoolsDoc(type: Javadoc, dependsOn: ['cleanReadtoolsDoc', classes]) {
```
Comment (magicDGS) [inline build.gradle]: This should not depend on `cleanReadtoolsDoc`, beacuse it removes the javadoc documentation if so.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        final CommandLineProgramProperties properties = currentWorkUnit.getCommandLineProperties();
+        if (properties == null) {
+            return super.getDescription(currentWorkUnit);
+        }
+        return properties.summary();
+    }
+
+    @Override
+    public String getDestinationFilename(final DocWorkUnit workUnit) {
+        // TODO: should use getDoclet().getOutputFileExtension() but requires that it is public
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Open an issue, and also a PR/issue in Barclay.

File: src/main/java/org/magicdgs/readtools/tools/trimming/TrimReads.java [hunk 3]
```diff
@@ -80,6 +80,9 @@
                 + " in combination with the new ordering.\n"
                 + "\nFind more information in " + RTHelpConstants.DOCUMENTATION_PAGE,
         programGroup = RTManipulationProgramGroup.class)
+// TODO: this could not be documented because a known issue with the GATK filer plugin
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/tools/trimming/TrimReads.java]: We should add an issue to the repo.

File: src/main/java/org/magicdgs/readtools/utils/read/transformer/trimming/CutReadTrimmer.java [hunk 4]
```diff
@@ -24,15 +24,19 @@
...
 import org.broadinstitute.barclay.argparser.CommandLineException;
+import org.broadinstitute.barclay.help.DocumentedFeature;
 import org.broadinstitute.hellbender.utils.read.GATKRead;
 
 /**
  * Trimmer for cropping some bases in one or both sides of the read.
  *
  * @author Daniel Gomez-Sanchez (magicDGS)
  */
+@DocumentedFeature(groupName = RTHelpConstants.DOC_CAT_TRIMMERS, groupSummary = RTHelpConstants.DOC_CAT_TRIMMERS_SUMMARY, summary = "Trims the read certain number of bases at the end of the read.")
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/utils/read/transformer/trimming/CutReadTrimmer.java]: Better summary

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.File;
+import java.util.Map;
+
+/**
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class RTHelpDocletUnitTest extends RTBaseTest {
+
+    /** Folder where templates live. */
+    static final String DOCUMENTATION_TEMPLATES_FOLDER =
```
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Could be protected?

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            TestResourcesUtils.READTOOLS_MAIN_RESOURCES_DIRECTORY
+                    + "org/magicdgs/readtools/documentation";
+
+    static final RTHelpDoclet DOCLET = new RTHelpDoclet();
```
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Javadoc and could be protected?

File: build.gradle (hunk 1 above)
Comment (magicDGS) [inline build.gradle]: Done.

File: src/main/java/org/magicdgs/readtools/tools/trimming/TrimReads.java (hunk 3 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/tools/trimming/TrimReads.java]: Done (and updated comment).

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java (hunk 5 above)
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Done.

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java (hunk 6 above)
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Done.

File: src/main/java/org/magicdgs/readtools/utils/read/transformer/trimming/CutReadTrimmer.java (hunk 4 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/utils/read/transformer/trimming/CutReadTrimmer.java]: Done.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java (hunk 2 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Done (both).

**Comment by magicDGS** [comment]
The bugfix for the issue is in review in Barclay, so maybe we can wait til it is included.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        // TODO: currentWorkUnit.getCommandLineProperties().summary() - maybe this should be discouraged
+        final CommandLineProgramProperties properties = currentWorkUnit.getCommandLineProperties();
+        if (properties == null) {
+            return super.getDescription(currentWorkUnit);
+        }
+        return properties.summary();
+    }
+
+    /**
+     * Destination filename uses
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Change javadoc

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        final DocWorkUnit mockedUnit = Mockito.mock(DocWorkUnit.class);
+        Mockito.when(mockedUnit.getGroupName()).thenReturn("group_name");
+        Mockito.when(mockedUnit.getGroupSummary()).thenReturn("group_summary");
+        final Map<String, String> groupMap = DOCLET.getGroupMap(mockedUnit);
+        Assert.assertEquals(groupMap.get("id"), "group_name");
+        Assert.assertEquals(groupMap.get("name"), "group_name");
+        Assert.assertEquals(groupMap.get("summary"), "group_summary");
+        Assert.assertEquals(groupMap.get("supercat"), "other");
+    }
+
```
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Extra line.

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandlerUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        Assert.assertEquals(HANDLER.getTagFilterPrefix(), "");
+    }
+
+    @Test
+    public void testGetDestinationFilename() {
+        final DocWorkUnit mockedWorkUnit = Mockito.mock(DocWorkUnit.class);
+        Mockito.when(mockedWorkUnit.getName()).thenReturn("MyToolName");
+        Assert.assertEquals(HANDLER.getDestinationFilename(mockedWorkUnit), "MyToolName.md");
+    }
+
```
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandlerUnitTest.java]: Extra line.

Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandlerUnitTest.java]: Extra line.

File: build.gradle [hunk 10]
```diff
@@ -155,3 +179,58 @@ shadowJar {
...
+
+    options.addStringOption("settings-dir", readtoolsTemplates)
+    // TODO: uncomment the next line (https://github.com/magicDGS/ReadTools/issues/243) -> requires https://github.com/broadinstitute/barclay/pull/60
+    // options.addStringOption("index-file-extension", "yml")
+    options.addStringOption("output-file-extension", "md")
+    options.addStringOption("absolute-version", getVersion())
+    options.addStringOption("build-timestamp", new Date().format("dd-mm-yyyy hh:mm:ss"))
+    options.addStringOption("verbose")
+
+    // TODO: remove once teh index file extension is available in yml
```
Comment (magicDGS) [inline build.gradle]: Mention the issue #243.

File: src/main/java/org/magicdgs/readtools/RTHelpConstants.java [hunk 11]
```diff
@@ -46,21 +51,68 @@ private RTHelpConstants() {}
...
+    public static final String DOC_CAT_TRIMMERS = "Trimmers";
+    public static final String DOC_CAT_TRIMMERS_SUMMARY = "Algorithms used to trim the reads.";
+
+    // map each group name to a super-category
+    private static Map<String, String> groupToSuperCategory;
+    // initialize on demand the mapping between supercategories and group names
+    private static Map<String, String> getSuperCategoryMap() {
+        if (groupToSuperCategory == null) {
+            // TODO: initialize with GATK's and/or Picard's supercat map
+
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/RTHelpConstants.java]: remove empty line

File: src/main/java/org/magicdgs/readtools/RTHelpConstants.java [hunk 12]
```diff
@@ -46,21 +51,68 @@ private RTHelpConstants() {}
...
+            groupToSuperCategory.put(DOC_CAT_TRIMMERS, DOC_SUPERCAT_UTILITIES);
+            groupToSuperCategory.put(HelpConstants.DOC_CAT_READFILTERS, DOC_SUPERCAT_UTILITIES);
+        }
+        return groupToSuperCategory;
+    }
+
+    /** Supercategory not defined in the map. */
+    private static final String DOC_SUPERCAT_OTHER = "other";
+
+    public static String getSuperCategoryProperty(final String groupName) {
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/RTHelpConstants.java]: javadoc

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+/**
+ * The ReadTools Documentation work unit handler class that is the companion to
+ * {@link RTHelpDocWorkUnitHandler}.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class RTHelpDocWorkUnitHandler extends DefaultDocWorkUnitHandler {
+
+    /** Default generic template for the documentation work unit. */
+    public static final String DEFAULT_TEMPLATE_NAME = "generic.template.md";
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Change to "DEFAULT_TEMPLATE_NAME_PREFIX" and use "generic.template." instead.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+    public RTHelpDocWorkUnitHandler(HelpDoclet doclet) {
+        super(doclet);
+    }
+
+    /** Returns {@link #DEFAULT_TEMPLATE_NAME}. */
+    @Override
+    public String getTemplateName(final DocWorkUnit workUnit) {
+        return DEFAULT_TEMPLATE_NAME;
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Use DEFAULT_TEMPLATE_NAME_PREFIX + outputExtension instead (see comment above)

Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Although the extension could not be used yet (see https://github.com/magicDGS/ReadTools/issues/240)

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+ */
+public class RTHelpDoclet extends HelpDoclet {
+
+    /** Markdown extension is our default. */
+    public static final String MARKDOWN_OUTPUT_FILE_EXTENSION = ".md";
+
+    /**
+     * Generic index template name for ReadTools documentation.
+     */
+    public static final String INDEX_TEMPLATE_NAME = "generic.index.template.yml";
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java]: Change also for INDEX_TEMPLATE_PREFIX ("generic.index.template.")

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+    /** Constructor with our {@link #MARKDOWN_OUTPUT_FILE_EXTENSION}. */
+    public RTHelpDoclet() {
+        super();
+        // default extension is Markdown
+        outputFileExtension = MARKDOWN_OUTPUT_FILE_EXTENSION;
```
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java]: TODO: override the indexFileExtension

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    @Test
+    public void testTrimmersDocGenDontBlowUp() throws Exception {
+        // run javadoc with our custom doclet to check if everything is working
+        // only test if the output folder is not empty
+        final File outputDir = createTestTempDir("DocGenTest");
+        final List<String> docArgList = Arrays.asList(
+                "-build-timestamp", "2016/01/01 01:01:01",      // dummy, constant timestamp
+                "-absolute-version", "11.1",                    // dummy version
+                "-output-file-extension", "md",                 // testing markdown output
+                // TODO: this should be uncommented at some point
```
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Add issue number

File: build.gradle (hunk 10 above)
Comment (magicDGS) [inline build.gradle]: Done

File: src/main/java/org/magicdgs/readtools/RTHelpConstants.java (hunk 11 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/RTHelpConstants.java]: Done.

File: src/main/java/org/magicdgs/readtools/RTHelpConstants.java (hunk 12 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/RTHelpConstants.java]: Done

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java (hunk 13 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Done.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java (hunk 14 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDocWorkUnitHandler.java]: Done.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java (hunk 15 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java]: Done.

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java (hunk 16 above)
Comment (magicDGS) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java]: Added TODO

File: src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java (hunk 17 above)
Comment (magicDGS) [inline src/test/java/org/magicdgs/readtools/documentation/RTHelpDocletUnitTest.java]: Done

File: src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+                new RTHelpDocWorkUnitHandler(this),
+                documentedFeature,
+                commmandLineProgramProperties,
+                classDoc,
+                clazz);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
```
Comment (sputnikci) [inline src/main/java/org/magicdgs/readtools/documentation/RTHelpDoclet.java]: \[PMD\] INFO: Ternary operators that can be simplified with || or &&

**Pull request merged**